File: Script/WebSocket/EventListenerList/AddWaiting.php
<?php
require_once __DIR__ . "/../EventListener.php";
require_once __DIR__ . '/../../Util.php'; 
require_once __DIR__ . '/../WebSocket.php'; 
require_once __DIR__ . '/../WebSocketMysql.php'; 
require_once Util::CONFIG_PATH;

use AWS\CRT\Internal\Encoding;
use Ratchet\ConnectionInterface;

class AddWaiting implements EventListener{
    public function OnCall(ConnectionInterface $from, $json_data) {

        // 전달받은 user_id
        $waitingUserName = $json_data[KEY_WAITING_USER_NAME]; // 친구로 추가할 ID
        $userId = $json_data[KEY_USER_ID]; // 현재 사용자의 ID
        $username = $json_data[KEY_USERNAME]; // 현재 사용자의 ID

        $response = array();
        $mysqli = Util::Instance()->getMysqli();


        $row = WebSocketMysql::Instance()->getUserByUsername($waitingUserName);
        if ($row) {
            $waiting = json_decode($row['waiting'], true);
            $friends = json_decode($row['friends'], true);

            if (!is_array($waiting)) {
                $waiting = [];
            }

            //데이터가 존제하면 추가하지않고 실패 반환
            if (in_array($userId, $waiting) || in_array($userId, $friends)) {
                $response = array(
                    TYPE => TYPE_ADD_WAITING,
                    STATUS => STATUS_SUCCESS . "_0",
                    KEY_DATA => "user id already added"
                );
            } else {
                $waiting[] = $userId;

                // 업데이트된 waiting 배열을 다시 JSON으로 인코딩
                $newWaitingJson = json_encode($waiting);

                // 데이터베이스에 업데이트
                $stmt = $mysqli->prepare("UPDATE users SET waiting = ? WHERE username = ?");
                $stmt->bind_param('ss', $newWaitingJson, $waitingUserName);

                if ($stmt->execute() == true) {
                    $data = array(
                        TYPE => TYPE_ADD_WAITING,
                        KEY_USER_ID  => $userId,
                        KEY_USERNAME => $username
                    );
                    WebSocket::Instance()->sendData($row["id"], json_encode($data));

                    $response = array(
                        TYPE => TYPE_ADD_WAITING,
                        STATUS => STATUS_SUCCESS,
                        KEY_DATA => "success"
                    );
                } else {
                    $response = array(
                        TYPE => TYPE_ADD_WAITING,
                        STATUS => STATUS_ERROR,
                        KEY_DATA => "fail to database update"
                    );
                }
            }
        } else {
            $response = array(
                TYPE => TYPE_ADD_WAITING,
                STATUS => STATUS_ERROR,
                KEY_DATA => "fail to find user"
            );
        }
        echo $userId . json_encode($response) . "\n";
        $from->send(json_encode($response));
        $mysqli->close();
    }
} 
?>

File: SocketScript/java/event/listener/AddWaiting.java
package event.listener;

import org.example.MysqlManager;
import org.example.SocketConnection;
import org.example.SocketEventListener;
import org.example.Util;
import org.json.JSONArray;
import org.json.JSONObject;

import java.nio.channels.SocketChannel;
import java.util.Map;

public class AddWaiting implements SocketEventListener.EventListener {
    @Override
    public void run(SocketChannel channel, JSONObject jsonObject) {

        String waitingUserName = jsonObject.getString(SocketEventListener.eKey.WAITING_USER_NAME.toString());
        int userId = jsonObject.getInt(SocketEventListener.eKey.USER_ID.toString());
        String username = jsonObject.getString(SocketEventListener.eKey.USERNAME.toString());

        Map<String, Object> waitingUser = MysqlManager.Instance().getUserByUsername(waitingUserName);
        if (waitingUser != null) {
            JSONArray waiting = new JSONArray(waitingUser.get("waiting").toString());
            JSONArray friends = new JSONArray(waitingUser.get("friends").toString());

            if (Util.isInArray(waiting, userId) || Util.isInArray(friends, userId)) {

                SocketConnection.trySendMessage(channel, new JSONObject()
                        .put(SocketEventListener.eKey.TYPE.toString(), SocketEventListener.eType.ADD_WAITING.toString())
                        .put(SocketEventListener.eKey.STATUS.toString(), "success_0")
                        .put(SocketEventListener.eKey.DATA.toString(), "user id already added")
                        .toString());
                return;
            }

            waiting.put(userId);

            if (MysqlManager.Instance().updateUsersWaitingByUsername(waiting, waitingUserName)) {
                SocketConnection.LOG(waitingUser.get("id").toString());

                SocketChannel socketChannel;
                if((socketChannel = SocketConnection.Instance().userIdMap.get(((Long)waitingUser.get("id")).intValue())) != null){
                    SocketConnection.trySendMessage(socketChannel, new JSONObject()
                            .put(SocketEventListener.eKey.TYPE.toString(), SocketEventListener.eType.ADD_WAITING.toString())
                            .put(SocketEventListener.eKey.USER_ID.toString(), userId)
                            .put(SocketEventListener.eKey.USERNAME.toString(), username)
                            .toString());
                }

                SocketConnection.trySendMessage(channel, new JSONObject()
                        .put(SocketEventListener.eKey.TYPE.toString(), SocketEventListener.eType.ADD_WAITING.toString())
                        .put(SocketEventListener.eKey.STATUS.toString(), "success")
                        .put(SocketEventListener.eKey.DATA.toString(), "success")
                        .toString());

            } else {
                SocketConnection.trySendMessage(channel, new JSONObject()
                        .put(SocketEventListener.eKey.TYPE.toString(), SocketEventListener.eType.ADD_WAITING.toString())
                        .put(SocketEventListener.eKey.STATUS.toString(), "error")
                        .put(SocketEventListener.eKey.DATA.toString(), "fail to db update")
                        .toString());

            }
        } else {
            SocketConnection.trySendMessage(channel, new JSONObject()
                    .put(SocketEventListener.eKey.TYPE.toString(), SocketEventListener.eType.ADD_WAITING.toString())
                    .put(SocketEventListener.eKey.STATUS.toString(), "error")
                    .put(SocketEventListener.eKey.DATA.toString(), "fail to find user")
                    .toString());

        }
    }
}
